src/ik_memetic.cpp

```diff
@@ -0,0 +1,239 @@
+#include <pick_ik/goal.hpp>
+#include <pick_ik/ik_gradient.hpp>
+#include <pick_ik/ik_memetic.hpp>
+#include <pick_ik/robot.hpp>
+
+#include <algorithm>
+#include <chrono>
+#include <cmath>
+#include <fmt/core.h>
+#include <optional>
+#include <vector>
+
+namespace pick_ik {
+
+MemeticIk MemeticIk::from(std::vector<double> const& initial_guess, CostFn const& cost_fn) {
+    return MemeticIk{std::vector<double>(initial_guess.size(), 0.0), cost_fn(initial_guess)};
+}
+
+void MemeticIk::computeExtinctions() {
+    double min_fitness = population_.front().fitness;
+    double max_fitness = population_.back().fitness;
+    for (size_t i = 0; i < population_count_; ++i) {
+        double grading = static_cast<double>(i) / static_cast<double>(population_count_ - 1);
+        population_[i].extinction =
+            (population_[i].fitness + min_fitness * (grading - 1)) / max_fitness;
+    }
+}
+
+void MemeticIk::gradientDescent(size_t const i, Robot const& robot, CostFn const& cost_fn) {
+    auto& individual = population_[i];
+    auto local_ik = GradientIk::from(individual.genes, cost_fn);
+    auto constexpr timeout_local = 0.005;
+    auto constexpr max_iters_local = 25;
+    auto const timeout_point_local =
+        std::chrono::system_clock::now() + std::chrono::duration<double>(timeout_local);
+
+    int iter = 0;
+    while (std::chrono::system_clock::now() < timeout_point_local && iter < max_iters_local) {
+        if (!step(local_ik, robot, cost_fn)) {
+            break;
+        }
+        iter++;
+    }
+
+    individual.genes = local_ik.best;
+    individual.fitness = cost_fn(individual.genes);
+    individual.gradient = local_ik.gradient;
+}
+
+void MemeticIk::initPopulation(Robot const& robot,
+                               CostFn const& cost_fn,
+                               std::vector<double> const& initial_guess) {
+    std::vector<double> const zero_grad(robot.variables.size(), 0.0);
+    population_.resize(population_count_);
+    for (size_t i = 0; i < elite_count_; ++i) {
+        auto genotype = initial_guess;
+        if (i > 0) {
+            for (size_t j_idx = 0; j_idx < robot.variables.size(); ++j_idx) {
+                auto const& var = robot.variables[j_idx];
+                genotype[j_idx] = uniform_dist_(gen_) * var.span + var.min;
+            }
+        }
+        population_[i] = Individual{genotype, cost_fn(genotype), 1.0, zero_grad};
+    }
+
+    // Initialize children to some dummy values that will be overwritten.
+    for (size_t i = elite_count_; i < population_count_; ++i) {
+        population_[i] = Individual{initial_guess, 0.0, 1.0, zero_grad};
+    }
+
+    computeExtinctions();
+}
+
+void MemeticIk::reproduce(Robot const& robot, CostFn const& cost_fn) {
+    std::vector<Individual*> pool;
```

Comment:
vector of raw pointers looks weird to me... do you know why it was done this way?

Reply:
I think it was done to avoid copies ... but if I just keep it as a vector of `Individual`, then `emplace_back` should be safe, right?

Reply:
I just thought it was weird, doesn't mean you need something more complex if this solves the problem. So long as it is contained in the function I'm ok with it.

Reply:
I think I'm actually going to abstract this whole "mating pool" construct a little more so it's more contained than it is right now. Thanks!